### DISCUSSION TITLE
Get the value of a MenuItem when callback function is executed

### DISCUSSION THREAD

**Discussion by agreen72** [forntoh/LcdMenu#14] [Q&A] [answered]
Hello Mr. Thomas,

I am doing a lot of research and learning alot about c++ in order to utilize your library in a project.  I am trying to use an ItemInput() with a callback function f(), in order to manipulate the String value parameter of the ItemInput().  But I am not sure how to get and set the value parameter of the ItemInput() object, within the callback function f().

My goal is to use a joystick to cycle each character in the value String to whatever the user wishes from the ASCII table, and then update the contents of the value parameter of the ItemInput() object.  This value will later be saved as a setting in the EEPROM of a device, to be called as a parameter for a function that will be recording measurements from sensor over a period of time.

I can currently explicitly grab the value using something like 'String x = mainMenu[4].value;'  But I would like something more generic so that I can reuse the function on any ItemInput().value that calls the callback function f().

How would I pass a reference of the ItemInput() object to the callback function()? Or more directly, how would I get and set the value String parameter of the ItemInput() object, within the callback function? I would imagine it may be passing an implicit 'this' reference, that I could then get or set the value parameter with a this->value call.

Any answers or resources you have in accomplishing this would be greatly appreciated.  Thank you.

I should note I also adapted the inputs for your library to come from a joystick with a press button. I have created five input commands to be interpreted from the joystick, up down left right and press, and debounced them. I've only used 4 so far as you only have four callable functions for menu navigation.

I assume you designed the ItemInput() value Strings to be set by your keypad input device.

p.s.

I noticed that the newer versions of LiquidCrystal_I2C.h have deprecated the init() function and replaced it with a begin() function that does something similar.  In my project I have manually replaced init() with begin() for the one occurence it has in your LcdMenu.h . 

Also in my research, I've learned that C++11 deprecated the conversion of Strings to char* types, and your example sketches,  for the name parameters of the menuItem objects, pass Strings to the menuItem which is expecting char*.   This comes up with a warning, but still compiles.  Answers I've seen on stack exchange say this is left in for backward compatibility but should not be used, as it may lead to undefined behavior in the compiler.  The easy solution is to declare the names as char* arrays outside the mainMenu declaration, and reference them for the name parameters of the menuItem() objects, instead of typing them in directly as Strings.

Also a getTop() function on LcdMenu would be greatly appreciated when selecting digits to manipulate.  The actual row to display a blinking cursor is at  (myLcdMenu.getCursorPosition() - myLcdMenu.top)

**Comment by agreen72** [comment]
I have found that you can use your LcdMenu object and it's setText() member function to update the String value parameter, from within the callback function.  This is because the library remembers the current menu position when calling the enter() function on a menuItem.  

Also to get the String value member parameter for the current menuItem there are 2 functions buried in the library I just dug up.

`MenuItem* objPtr = menu.getItemAt(menu.getCursorPosition());
  Serial.println(objPtr -> value);`

myLcdMenu.getItemAt(uint8_t position) to get &currentMenuTable[position]
myLcdMenu.getCursorPosition() to get the uint8_t currentCursorPosition to feed to the above line
it will return a MenuItem pointer you declare with      MenuItem* name

Glad these were imbedded in there, but I really think  you ought to update and flesh out your examples for each of these common abilities.  Knowing I could have done this from the start would have saved me about a day and a half of head scratching.  Although I did learn all about pointers and callback functions in c++ in the meantime.

**Reply by forntoh** [reply]
Oh... I pushed you to learn more 😉
Nevermind, I will update the [Wiki](https://github.com/forntoh/LcdMenu/wiki/Example-03-Callback)

**Reply by forntoh** [reply]
I am currently writing some code that will make this way easier,

How? 🤔

The value of the `ItemInput` or the toggle state will be passed directly to the call back function 🤩,

```rust
//... stuff
ItemInput("Name","", saveName)
//... more stuff
void saveName(String name) {
    // 🤗 do stuff with name
}
```

See #36

**Comment by agreen72** [comment]
I made a callback function that lets me edit an input with only a joystick.  Incrementally 1 character at a time. The blinking cursor highlights the digit be altered.  Without getTop the blinking cursor sometimes appears on the wrong row although still in the correct column.  Left right on joy select digit up down increments or decrement the digit, pressing in the button setTexts the item input menu item to update the value member variable, and redraw screen.

Sent from my iPhone

On Mar 16, 2021, at 8:21 AM, Forntoh Thomas ***@***.***> wrote:

﻿

Also a getTop() function on LcdMenu would be greatly appreciated when selecting digits to manipulate. The actual row to display a
blinking cursor is at (myLcdMenu.getCursorPosition() - myLcdMenu.top)

@agreen72<https://github.com/agreen72> can you explain to me when you would need to use getTop?

**Reply by forntoh** [reply]
Okay, I understand

New #16

**Comment by forntoh** [comment, marked as answer]
Beginning in [version 2.0.0](https://github.com/forntoh/LcdMenu/releases/tag/2.0.0), the value of the `ItemInput`, the position of `ItemList` or the toggle state is passed directly to the callback function,

```rust
//... stuff
//
ItemInput("Name","", saveName)
//
//... more stuff
//
void saveName(String name) {
    // 🤗 do stuff with name
}
```

**Reply by Collin7** [reply]
Hi @forntoh 

Is it possible to do this for ItemCommand as well.
My thinking is that instead of creating multiple callbacks for each ItemCommand, I could have a single one that can be used for all Items and with a switch case dependent on the Item Name, I could, in my case publish a command to the correct MQTT topic.

Hope this makes sense.

Thanks.

**Reply by forntoh** [reply]
You can definitely use the same callback for multiple menu items, you just need to get the current (active) item from your menu in your callback to know which one called the function.

`menu[menu.getCursorPosition()]` will give you the active item

**Reply by Collin7** [reply]
Love the library by the way.
Thank you. that is helpful.
Is it also possible to get the parent menu item from a submenu Item?
So when in a submenu, to know which parent menuItem was selected?

**Reply by forntoh** [reply]
Thanks 😊 I'm happy that you love what I created. A star will be helpful.

YES, you can get the parent menu. You bind a submenu to its parent in the `ItemHeader` of the submenu, right!
So, getting the item at index = 0 from the menu will give you the `ItemHeader`, from there you can get its parent menu. The naming is kind of messed up for your use case... calling [getSubmenu()](https://github.com/forntoh/LcdMenu/blob/master/src/MenuItem.h#L79) on an `ItemHeader` returns the parent menu if the submenu where the `ItemHeader` is.

I hope that you understand.

**Reply by Collin7** [reply]
Starred it :) and shared with my other Dev friends to use.
I was 60% done writing my own, and then came across this.
Goes along way with my home automation stuff :)

That makes sense thank you.  I should have noticed that actually.
Thanks

**Reply by Collin7** [reply]
sorry @forntoh 
I was trying to implement this now but realized the missing piece I was looking for.
So I understand what you are saying, I get get the parent menu from the submenu header item,
but can I get the actual item that was selected in the parent menu?

So lets say my main menu is
SubMenuItem1
SubMenuItem2
SubMenuItem3
SubMenuItem4

then if I select SubMenuItem3 and go to that subMenu that has might have 
ItemCommand1
ItemCommand2

If I select ItemCommand2 -> In the callback can I some how determine that The parent selected was SubMenuItem3?

**Reply by forntoh** [reply]
Following my previous suggestion, you will get a reference to the parent menu (which is a pointer to an array), because a menu is just a collection of items which in c++ is a pointer reference (not an object) so it's not easy to know which is which. I guess the best option for you is to use names (labels) that are unique so that you won't need to know the parent, you can know which action to perform from the label 🤷‍♂️
